[Android / Ocaml 5] undefined symbol: pthread_getaffinity_np

## Environment

### OPAM Packages Installed
<details>

```
/data/data/com.termux/files/usr/include % opam list                                         # Packages matching: installed
# Name        # Installed # Synopsis
base-bigarray base
base-threads  base
base-unix     base
cmdliner      1.0.4       Declarative definiti
ocaml         4.07.1      The OCaml compiler (
ocaml-config  1           OCaml Switch Configu
ocaml-system  4.07.1      The OCaml compiler (
ocamlfind     1.9.1       A library manager fo
seq           base        Compatibility packag
/data/data/com.termux/files/usr/include %
```

</details>

### Details
<details><summary> `neofetch --off` </summary>

```
u0_a248@localhost
-----------------                             OS: Android 11 aarch64
Host: Redmi 21091116AC
Kernel: 4.14.186-perf-gd65dbc980d48
Uptime: 1 day, 14 hours, 51 mins
Packages: 300 (dpkg), 1 (pkg)
Shell: zsh 5.9
CPU: MT6833P (8) @ 2.000GHz                   Memory: 2753MiB / 5635MiB
```
</details>

### uname -a
<details>

```

Linux localhost 4.14.186-perf-gd65dbc980d48 #1 SMP PREEMPT Fri Oct 28 22:22:47 CST 2022 aarch64 Android

```

</details>

### OPAM Version

2.0.3

***

## Error Message

<details><summary> `opam switch install 5.0.0` Error Message </summary>

```

(Skipping...)

# 2 warnings generated.
# gcc -O2 -fno-strict-aliasing -fwrapv -pthread -g -Wall -fno-common -ffunction-sections  -Wl,-E -landroid-shmem -o runtime/ocamlrun runtime/prims.o runtime/libcamlrun.a -lm -lexecinfo  -lpthread
# ld.lld: error: undefined symbol: pthread_getaffinity_np
# >>> referenced by domain.c:1795 (runtime/domain.c:1795)
# >>>               domain.b.o:(caml_recommended_domain_count) in archive runtime/libcamlrun.a
# clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
# make[2]: *** [Makefile:822: runtime/ocamlrun] Error 1
# make[2]: Leaving directory '/data/data/com.termux/files/home/.opam/5.0.0/.opam-switch/build/ocaml-base-compiler.5.0.0'

(Skipping...)

```

</details>

### $(grep -r pthread_get)
<details>

```bash

/data/data/com.termux/files/usr/include % grep -r pthread_get
npth.h:#define npth_getspecific pthread_getspecific
npth.h:   pthread_getspecific are standardized as async-signal-safe (what a
npth.h:   pthread_self and pthread_getspecific is probably portable
pthread.h:int pthread_getattr_np(pthread_t __pthread, pthread_attr_t* __attr);              pthread.h:int pthread_getcpuclockid(pthread_t __pthread, clockid_t* __clock);
pthread.h:int pthread_getschedparam(pthread_t __pthread, int* __policy, struct sched_param* __param);                                     pthread.h:void* pthread_getspecific(pthread_key_t __key);
pthread.h:pid_t pthread_gettid_np(pthread_t __pthread) __INTRODUCED_IN(21);                 pthread.h:int pthread_getname_np(pthread_t __pthread, char* __buf, size_t __n) __INTRODUCED_IN(26);
c++/v1/__threading_support:  return pthread_getspecific(__key);
bits/threads_inlines.h:  return pthread_getspecific(__key);                                 python3.11/pyconfig.h:/* Define to 1 if you have the `pthread_getcpuclockid' function. */
python3.11/cpython/pthread_stubs.h:PyAPI_FUNC(void *) pthread_getspecific(pthread_key_t key);
```
</details>

***
## \`opam config\` report

```

# opam config report
# opam-version      2.0.3
# self-upgrade      no
# system            arch=arm64 os=linux os-distribution=android os-version=11
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              1
# repositories      1 (http) (default repo at 25bba47d)
# pinned            0
# current-switch    default

```

***

## Note

- Native Termux, not proot-distro linux
- Ocaml & OPAM pre-compiled binary installed via the famous Termux 3rd party repo 'its-pointless'
- LDFLAGS="-landroid-shmem"
- With above configurations `opam switch install 4.14.1` was successful, now `opam switch list` looks like this:
```

#  switch   compiler
          description                         →  4.14.1   ocaml-base-compiler.4.14.1
          4.14.1                                 default  ocaml-system.4.07.1
          default
```

> Anji (2023/03/14 03:15 -0700):
> > > I'm somewhat puzzled, since two hours ago I posted two comments on this issue. I'm hoping I didn't post them on another issue by mistake... > > It turns out I'm going slightly mad - I posted the patch on [[ocaml/opam#5477](https://github.com/ocaml/opam/issues/5477) (comment)]([ocaml/opam#5477 (comment)](https://github.com/ocaml/opam/issues/5477#issuecomment-1465129163)). Sorry for confusing you mate! My bad. I should not have created multiple issues. I'm really sorry.
> Did you give David's patch a try? Does it work for you?

Umm, first of all, I'm a noob. And no, it did not work for me. This is how I applied the patch: I git clone OCaml repo, find `configure.ac`, manually modified it, and then `./configure --prefix=$PREFIX`, here are some error messages complaining about `pthread` functions.

---

<details>
 <summary> 
The Error Message from `make` </summary>

```
(Skipping...)

CCDEPS .dep/runtime/frame_descriptors.npic.d
  CCDEPS .dep/runtime/startup_nat.npic.d
  CCDEPS .dep/runtime/signals_nat.npic.d
make coldstart
make[1]: Entering directory '/data/data/com.termux/files/home/ocaml/ocaml'
make runtime-all
make[2]: Entering directory '/data/data/com.termux/files/home/ocaml/ocaml'
  GEN runtime/ld.conf
  CC runtime/addrmap.b.o
  CC runtime/afl.b.o
  CC runtime/alloc.b.o
  CC runtime/array.b.o
  CC runtime/backtrace.b.o
  CC runtime/bigarray.b.o
  CC runtime/callback.b.o
  CC runtime/codefrag.b.o
  CC runtime/compare.b.o
  CC runtime/custom.b.o
  CC runtime/debugger.b.o
  CC runtime/domain.b.o
runtime/domain.c:1870:7: error: call to undeclared function 'pthread_getaffinity_np'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
  if (pthread_getaffinity_np(pthread_self(), sizeof(cpuset), &cpuset) == 0)
      ^
runtime/domain.c:1870:7: note: did you mean 'pthread_getattr_np'?
/data/data/com.termux/files/usr/include/pthread.h:188:5: note: 'pthread_getattr_np' declared here
int pthread_getattr_np(pthread_t __pthread, pthread_attr_t* __attr);
    ^
1 error generated.
make[2]: *** [Makefile:969: runtime/domain.b.o] Error 1
make[2]: Leaving directory '/data/data/com.termux/files/home/ocaml/ocaml'
make[1]: *** [Makefile:209: coldstart] Error 2
make[1]: Leaving directory '/data/data/com.termux/files/home/ocaml/ocaml'
make: *** [Makefile:375: world.opt] Error 2

```

</details>

---

Excuse me for being a noob, I barely have any CS knowledge and have only used Python to make some GUI applications.

Btw, I successfully installed OCaml 5 inside Termux's proot-distro alpine linux instead of native Termux before I rooted my device.

> When you modify `configure.ac`, there needs to be a build step to derive the new `configure` script from it. `make configure` should do the job, assuming `autoconf` is installed on your machine.

Progress! `runtime/domain*` was compiled, but new error message (GC related?) appeared

```
(Skipping...)

  CC runtime/debugger.b.o
  CC runtime/domain.b.o
  CC runtime/dynlink.b.o
  CC runtime/extern.b.o
  CC runtime/fiber.b.o
  CC runtime/finalise.b.o
  CC runtime/floats.b.o
  CC runtime/gc_ctrl.b.o
  CC runtime/gc_stats.b.o
  CC runtime/globroots.b.o
  CC runtime/hash.b.o
  CC runtime/intern.b.o
  CC runtime/ints.b.o
  CC runtime/io.b.o
  CC runtime/lexing.b.o
  CC runtime/lf_skiplist.b.o
  CC runtime/main.b.o
  CC runtime/major_gc.b.o
runtime/major_gc.c:943:9: error: 'PAGE_MASK' macro redefined [-Werror,-Wmacro-redefined]
#define PAGE_MASK (~(uintnat)(BITS_PER_WORD-1))
        ^
/data/data/com.termux/files/usr/include/sys/user.h:39:9: note: previous definition is here
#define PAGE_MASK (~(PAGE_SIZE - 1))
        ^
1 error generated.
make[2]: *** [Makefile:969: runtime/major_gc.b.o] Error 1
make[2]: Leaving directory '/data/data/com.termux/files/home/ocaml/ocaml'
make[1]: *** [Makefile:209: coldstart] Error 2
make[1]: Leaving directory '/data/data/com.termux/files/home/ocaml/ocaml'
make: *** [Makefile:375: world.opt] Error 2
~/ocaml/ocaml $
```

Looks like the macro name `PAGE_MASK` conflicted with some header file from `sys/`, not sure what this `user.h` does

> You want to pass `--disable-warn-error` to `./configure` to disable this kind of inessential errors.
> 
> (We should investigate why one of the Android headers that we include also defines `PAGE_MASK`, or prefix the macro, or disable the warning.)

Thanks a lot, I retried with `LDFLAGS="-landroid-shmem" make` and now I'm figuring out what to do with this `libandroid-shmat` error:

```
  OCAMLC options.cmi
  OCAMLC options.cmo
  OCAMLC main.cmi
  OCAMLC main.cmo
  LINKC ocamltest
ld.lld: error: undefined symbol: libandroid_shmat
>>> referenced by afl.c:101 (runtime/afl.c:101)
>>>               afl.b.o:(caml_setup_afl) in archive ../stdlib/libcamlrun.a
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
File "_none_", line 1:
Error: Error while building custom runtime system
make[3]: *** [Makefile:138: ocamltest] Error 2
make[3]: Leaving directory '/data/data/com.termux/files/home/ocaml/ocaml/ocamltest'
make[2]: *** [Makefile:1223: ocamltest] Error 2
make[2]: Leaving directory '/data/data/com.termux/files/home/ocaml/ocaml'
make[1]: *** [Makefile:303: opt.opt] Error 2
make[1]: Leaving directory '/data/data/com.termux/files/home/ocaml/ocaml'
make: *** [Makefile:376: world.opt] Error 2

```

**I ran into this `libandroid-shmat` problem before, and usually when I add `LDFLAGS="-landroid-shmem"`, the problem goes away**; this time I'm not sure what to do with it,

```
~/ocaml/ocaml $ apt search libandroid
Sorting... Done
Full Text Search... Done
libandroid-complex-math/stable,now 0.1 aarch64 [installed]
  A shared library providing libm complex math functions

libandroid-complex-math-static/stable,now 0.1 aarch64 [installed]
  Static libraries for libandroid-complex-math

libandroid-execinfo/stable,now 0.1-1 aarch64 [installed]
  Shared library for the backtrace system function

libandroid-execinfo-static/stable,now 0.1-1 aarch64 [installed]
  Static libraries for libandroid-execinfo

libandroid-glob/stable,now 0.6-2 aarch64 [installed]
  Shared library for the glob(3) system function

libandroid-glob-static/stable,now 0.6-2 aarch64 [installed]
  Static libraries for libandroid-glob

libandroid-posix-semaphore/stable,now 0.1-3 aarch64 [installed]
  Shared library for the posix semaphore system function

libandroid-posix-semaphore-static/stable,now 0.1-3 aarch64 [installed]
  Static libraries for libandroid-posix-semaphore

libandroid-shmem/stable,now 0.4 aarch64 [installed]
  System V shared memory emulation on Android using ashmem

libandroid-shmem-static/stable,now 0.4 aarch64 [installed]
  Static libraries for libandroid-shmem

libandroid-spawn/stable,now 0.3 aarch64 [installed]
  Shared library for the posix_spawn system function

libandroid-spawn-static/stable,now 0.3 aarch64 [installed]
  Static libraries for libandroid-spawn

libandroid-support/stable,now 28-3 aarch64 [installed]
  Library extending the Android C library (Bionic) for additional multibyte, locale and math support

libandroid-support-static/stable,now 28-3 aarch64 [installed]
  Static libraries for libandroid-support

libandroid-sysv-semaphore/stable,now 0.1 aarch64 [installed]
  A shared library providing System V semaphores

libandroid-sysv-semaphore-static/stable,now 0.1 aarch64 [installed]
  Static libraries for libandroid-sysv-semaphore

libandroid-wordexp/stable,now 0.1 aarch64 [installed]
  Shared library for the wordexp(3) system function

libandroid-wordexp-static/stable,now 0.1 aarch64 [installed]
  Static libraries for libandroid-wordexp

```

I'll try to start from the very beginning with `make clean`

--- a/Changes
+++ b/Changes
@@ -770,6 +770,9 @@ OCaml 5.1.0
 - #12162: Fix miscompilation on amd64 backends involving integer overflows
   (Vincent Laviron and Greta Yorsh, review by Stefan Muenzel)
 
+- #12170: fix pthread_geaffinity_np configure check for android
+  (David Allsopp, review by Sébastien Hinderer)
+
 - #12178: Fix runtime events consumer poll function returning an invalid value
   instead of an OCaml integer value. (Lucas Pluvinage)
 

--- a/configure
+++ b/configure
@@ -17955,6 +17955,8 @@ fi
 
 
 ## pthread_getaffinity_np, args differ from GNU and BSD
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking arguments for pthread_getaffinity_np" >&5
+printf %s "checking arguments for pthread_getaffinity_np... " >&6; }
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 #define _GNU_SOURCE
@@ -17971,8 +17973,10 @@ cpu_set_t cs;
   return 0;
 }
 _ACEOF
-if ac_fn_c_try_compile "$LINENO"
+if ac_fn_c_try_link "$LINENO"
 then :
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: gnu" >&5
+printf "%s\n" "gnu" >&6; }
   printf "%s\n" "#define HAS_GNU_GETAFFINITY_NP 1" >>confdefs.h
 
 else $as_nop
@@ -17993,14 +17997,21 @@ cpuset_t cs;
   return 0;
 }
 _ACEOF
-if ac_fn_c_try_compile "$LINENO"
+if ac_fn_c_try_link "$LINENO"
 then :
-  printf "%s\n" "#define HAS_BSD_GETAFFINITY_NP 1" >>confdefs.h
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: BSD" >&5
+printf "%s\n" "BSD" >&6; }
+    printf "%s\n" "#define HAS_BSD_GETAFFINITY_NP 1" >>confdefs.h
 
+else $as_nop
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: pthread_getaffinity_np not found" >&5
+printf "%s\n" "pthread_getaffinity_np not found" >&6; }
 fi
-rm -f core conftest.err conftest.$ac_objext conftest.beam conftest.$ac_ext
+rm -f core conftest.err conftest.$ac_objext conftest.beam \
+    conftest$ac_exeext conftest.$ac_ext
 fi
-rm -f core conftest.err conftest.$ac_objext conftest.beam conftest.$ac_ext
+rm -f core conftest.err conftest.$ac_objext conftest.beam \
+    conftest$ac_exeext conftest.$ac_ext
 
 ## ZSTD compression library
 

--- a/configure.ac
+++ b/configure.ac
@@ -1924,7 +1924,8 @@ AC_CHECK_FUNC([ffs], [AC_DEFINE([HAS_FFS])])
 AC_CHECK_FUNC([_BitScanForward], [AC_DEFINE([HAS_BITSCANFORWARD])])
 
 ## pthread_getaffinity_np, args differ from GNU and BSD
-AC_COMPILE_IFELSE(
+AC_MSG_CHECKING([arguments for pthread_getaffinity_np])
+AC_LINK_IFELSE(
   [AC_LANG_PROGRAM(
     [[#define _GNU_SOURCE
       #include <sched.h>
@@ -1933,8 +1934,9 @@ AC_COMPILE_IFELSE(
       CPU_ZERO(&cs);
       CPU_COUNT(&cs);
       pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);]])],
-  [AC_DEFINE([HAS_GNU_GETAFFINITY_NP])],
-  [AC_COMPILE_IFELSE(
+  [AC_MSG_RESULT([gnu])
+  AC_DEFINE([HAS_GNU_GETAFFINITY_NP])],
+  [AC_LINK_IFELSE(
     [AC_LANG_PROGRAM(
       [[#include <pthread.h>
         #include <pthread_np.h>
@@ -1944,7 +1946,9 @@ AC_COMPILE_IFELSE(
         CPU_ZERO(&cs);
         CPU_COUNT(&cs);
         pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);]])],
-    [AC_DEFINE([HAS_BSD_GETAFFINITY_NP])])])
+    [AC_MSG_RESULT([BSD])
+    AC_DEFINE([HAS_BSD_GETAFFINITY_NP])],
+    [AC_MSG_RESULT([pthread_getaffinity_np not found])])])
 
 ## ZSTD compression library
 

